File: ANE/threshold_frame.m
function [frame_thred, noisy, final_thred] = threshold_frame(frame, thred_inten, target_area, unmasked)
    [Lxm,Lym] = size(frame);
    if ~exist('unmasked','var')
        unmasked = true(Lxm,Lym);
    end
    final_thred = thred_inten;
    frame_thred = frame > thred_inten;
    noisy = false;
    
    [L,nL] = bwlabel(frame_thred,4);
%     CC = bwconncomp(frame_thred,4);
%     nL = CC.NumObjects;
    if nL>=1
%         list_area_L = cellfun(@length, CC.PixelIdxList);
        % Find the thresholded region with the largest area
        list_area_L = zeros(nL,1);
        list_area_masked_L = zeros(nL,1);
        for ix = 1:Lxm
        for iy = 1:Lym
            val = L(ix,iy);
            if val
                list_area_L(val) = list_area_L(val)+1;
                list_area_masked_L(val) = list_area_masked_L(val)+unmasked(ix,iy);
            end
        end
        end
        [max_area] = max(list_area_L);
        if max_area < sum(frame_thred,'all')/2
            noisy = true;
        end
%         list_area_masked_L = cellfun(@(x) nnz(unmasked(x)), CC.PixelIdxList);
        [~,iLm] = max(list_area_masked_L);
%         core_region = false(Lxm,Lym);
%         core_region(CC.PixelIdxList{iLm})=1; 
        core_region = (L == iLm);
        
        % If the area of the core region is too small, iteratively lower the threshold
        if ~exist('target_area','var') || isempty(target_area) || max_area >= target_area
            frame_thred = core_region;
        else
            frame_thred_old = frame_thred;
            while max_area < target_area % && thred_inten > 0.1 
                final_thred = thred_inten;
                frame_thred_old = frame_thred;
                if thred_inten > 1
                    thred_inten = thred_inten * 0.9;
                else
                    thred_inten = thred_inten - 0.1;
                end
    %                 thred_inten = quantile(avg_frame_use, 1-mean(mask_sub,'all')/max_area*area(nn), 'all');
                frame_thred = frame > thred_inten;
%                 CC = bwconncomp(frame_thred,4);
%                 nL = CC.NumObjects;
%                 list_area_L = cellfun(@length, CC.PixelIdxList);
                [L,nL] = bwlabel(frame_thred,4);
                list_area_L = zeros(nL,1);
                for ix = 1:Lxm
                for iy = 1:Lym
                    val = L(ix,iy);
                    if val
                        list_area_L(val) = list_area_L(val)+1;
                    end
                end
                end
                [max_area] = max(list_area_L);
            end
            [L,nL] = bwlabel(frame_thred_old,4);
%             CC = bwconncomp(frame_thred_old,4);
%             nL = CC.NumObjects;

%             list_area_L_core = cellfun(@(x) nnz(core_region(x)), CC.PixelIdxList);
            list_area_L_core = zeros(nL,1);
            [xx,yy] = find(core_region);
            for ii = 1:length(xx)
                val = L(xx(ii),yy(ii));
                if val
                    list_area_L_core(val) = list_area_L_core(val)+1;
                end
            end
            [~,iL] = max(list_area_L_core);
            frame_thred = (L==iL);
%             frame_thred = false(Lxm,Lym);
%             frame_thred(CC.PixelIdxList{iL})=1; 
        end
    end
    
    % Fill the holes in the thresholded core region
%     CC = bwconncomp(~frame_thred,4);
%     nL0 = CC.NumObjects;
    [L0,nL0] = bwlabel(~frame_thred,4);
    if nL0>1
%         list_area_L0 = cellfun(@length, CC.PixelIdxList);
        list_area_L0 = zeros(nL0,1);
        for ix = 1:Lxm
        for iy = 1:Lym
            val = L0(ix,iy);
            if val
                list_area_L0(val) = list_area_L0(val)+1;
            end
        end
        end
        [~,iL0] = max(list_area_L0);
        frame_thred = (L0 ~= iL0);
%         frame_thred = true(Lxm,Lym);
%         frame_thred(CC.PixelIdxList{iL0})=false; 
    end

File: ANE/threshold_frame.py
import numpy as np
from scipy.ndimage import label
# import cv2


def threshold_frame(frame, thred_inten, target_area=None, unmasked=None):
    Lym, Lxm = frame.shape
    if unmasked is None:
        unmasked = np.ones((Lym, Lxm),'bool')

    final_thred = thred_inten
    frame_thred = frame > thred_inten
    noisy = False
    L, nL = label(frame_thred)
    # L, nL= cv2.connectedComponents(frame_thred, connectivity=4)

    if nL >= 1:
        # Find the thresholded region with the largest area
        list_area_L = np.zeros(nL)
        list_area_masked_L = np.zeros(nL)
        for iy in range(Lym):
            for ix in range(Lxm):
                val = L[iy, ix]
                if val:
                    list_area_L[val] = list_area_L[val] + 1
                    list_area_masked_L[val] = list_area_masked_L[val] + unmasked[iy, ix]
        max_area = list_area_L.max()
        if max_area < frame_thred.sum() / 2:
            noisy = True
        iLm = np.argmax(list_area_masked_L)
        core_region = (L == iLm)

        # If the area of the core region is too small, iteratively lower the threshold
        if (target_area is None) or (len(target_area) == 0) or (max_area >= target_area):
            frame_thred = core_region
        else:
            frame_thred_old = frame_thred
            while max_area < target_area:
                final_thred = thred_inten
                frame_thred_old = frame_thred
                if thred_inten > 1:
                    thred_inten = thred_inten * 0.9
                else:
                    thred_inten = thred_inten - 0.1
                frame_thred = frame > thred_inten
                L, nL = label(frame_thred)
                # L, nL= cv2.connectedComponents(frame_thred, connectivity=4)
                list_area_L = np.zeros(nL)
                for ix in range(Lxm):
                    for iy in range(Lym):
                        val = L[iy, ix]
                        if val:
                            list_area_L[val] = list_area_L[val] + 1
                max_area = list_area_L.max()

            L, nL = label(frame_thred_old)
            # L, nL= cv2.connectedComponents(frame_thred_old, connectivity=4)
            list_area_L_core = np.zeros(nL)
            yy, xx = core_region.nonzero()
            # for ii in range(len(xx)):
            #     val = L[yy[ii], xx[ii]]
            for (iy, ix) in np.zip(yy,xx):
                val = L[iy, ix]
                if val:
                    list_area_L_core[val] = list_area_L_core[val] + 1
            iL = np.argmax(list_area_L_core)
            frame_thred = (L == iL)

    # Fill the holes in the thresholded core region
    L0, nL0 = label(np.logical_not(frame_thred))
    # L0, nL0= cv2.connectedComponents(np.logical_not(frame_thred), connectivity=4)
    if nL0 > 1:
        list_area_L0 = np.zeros(nL0)
        for iy in range(Lym):
            for ix in range(Lxm):
                val = L0[iy, ix]
                if val:
                    list_area_L0[val] = list_area_L0[val] + 1
        iL0 = np.argmax(list_area_L0)
        frame_thred = (L0 != iL0)

    return frame_thred, noisy, final_thred
